### ISSUE TITLE
Document reasons for block listing domains with DNS to 127.0.0.1

### ISSUE THREAD

**Issue by xurizaemon** [easylist/easylist#16372]
Domains (including wildcarded domains) pointing to 127.0.0.1 are a commonly used developer feature. Some domains which use this are `*.lndo.site` and `*.xip.io`

Blocking DNS for those developer tools in this list affects developers who are using those tools (requiring them to override the domains) and affects the OOTB experience for users of tools like Lando, DDev etc.

In https://github.com/search?q=repo%3Aeasylist%2Feasylist+localhost&type=commits we can see a few commits which introduce blocking of domains that point to 127.0.0.1, but the reasoning for this isn't explicit. I don't see this listed in the README either, so it's hard to establish the criteria for blocking those domains. This means that comments such as ["still not acceptable"](https://github.com/easylist/easylist/pull/16353#issuecomment-1597039841) do not have clear criteria for what is or is not acceptable.

I could _guess_ at the reasoning, but it would be great to clarify the exact reasoning for blocking such domains, so that the tools / domains know the right way to engage this (eg, clarify their purpose, or document, or something else).

**Comment by oofdog** [comment]
@ryanbr Your input would be appreciated. As a heavy user of nip.io and xip.io, I and many other devs I know, have had their environments broken because ublock origin now blocks those domains. We've unblocked them locally, but we'd like to know why these were added to the list in the first place. What's the security concern?

**Comment by ryanbr** [comment]
Is a security issue, imagine if you're running a webserver on a site decided to access it from outside, whether to fingerprint you or act nefariously. There should be no reason why a third-party access localhost. But do tell me, why we should we trust sites accessing localhost.

**Comment by oofdog** [comment]
> Is a security issue, imagine if you're running a webserver on a site decided to access it from outside, whether to fingerprint you or act nefariously. There should be no reason why a third-party access localhost. But do tell me, why we should we trust sites accessing localhost.

I'm not seeing the issue here.

If a domain name resolves to localhost, this doesn't mean that the person who owns the domain name or the nameservers has access to the end user's machine.

Even if the user has a webserver running on their machine, all HTTP traffic would be sent to the local webserver, since those domains resolve to 127.0.0.1.

If there really is a security issue, could you document/explain it better, preferably with some kind of diagram?

**Comment by ryanbr** [comment]
Give me a testable, valid example where I want a site to check for my localhost. Trusting a third-party "to do the right thing" when accessing 127.0.0.1 resources.

For reference, we already block localhost on [specific domains](https://github.com/easylist/easylist/blob/master/easyprivacy/easyprivacy_specific.txt#LL1285C1-L1286C1):



`||127.0.0.1^$third-party,domain=53.com|ameriprise.com|beachbody.com|chick-fil-a.com|citi.com|ebay.at|ebay.be|ebay.ca|ebay.ch|ebay.cn|ebay.co.uk|ebay.com|ebay.com.au|ebay.com.hk|ebay.com.my|ebay.com.sg|ebay.de|ebay.es|ebay.fr|ebay.ie|ebay.it|ebay.nl|ebay.ph|ebay.pl|equifax.ca|equifax.com|globo.com|gumtree.com|kleinanzeigen.de|lendup.com|mbna.ca|rusneb.ru|sciencedirect.com|sky.com|spectrum.net|td.com|tiaa.org|vedacheck.com|wepay.com|whatleaks.com`

**Comment by almunnings** [comment]
What does it matter if a 3rd party accesses 127.0.0.1? They're accessing their own computer? Whats the missing link here?

**Comment by xurizaemon** [comment]
> Give me a testable, valid example where I want a site to check for my localhost.

This is not the use case, nor (AFAIK) possible.

A third party may be able to trigger a request to my localhost (this is already possible without DNS: `<img src="https://127.0.0.1/some/resource">`) but they cannot observe the outcome of that request.

**Comment by trice-ardc** [comment]
> Give me a testable, valid example where I want a site to check for my localhost. Trusting a third-party "to do the right thing" when accessing 127.0.0.1 resources.

What is a testable example where someone can attack my local machine by pointing some dns at 127.0.0.1?

**Comment by mosajjal** [comment]
> Is a security issue, imagine if you're running a webserver on a site decided to access it from outside, whether to fingerprint you or act nefariously. There should be no reason why a third-party access localhost. But do tell me, why we should we trust sites accessing localhost.

agree with this. you can easily port scan a host and identify OS/version based on the responses. for example `127.0.0.1:49152` is usually open in Windows systems. Linux systems usually run ntp and ssh. not to mention Tor being present (`localhost:9050`). Developers can add a random host to `/etc/hosts` and use that for testing surely :)

**Comment by xurizaemon** [comment]
@mosajjal as per your example, that vector does not require DNS; blocking a domain such as those mentioned here does not add protection to the users. Speaking for myself, I use tools such as DNS specifically to avoid editing `/etc/hosts`.

**Comment by twitchyliquid64** [comment]
> Reasons to block DNS to 127.0.0.1

From a web context, yall want to be googling DNS rebinding attacks. They are particularly nasty when javascript is able to probe something listening on the local machine and potentially escalate privileges.

I'm not familiar with easylist and what it achieves and what it aims, but simply wanted to jump in to suggest "DNS rebinding attacks" is part of your reading.

**Comment by oofdog** [comment]
> agree with this. you can easily port scan a host and identify OS/version based on the responses. for example `127.0.0.1:49152` is usually open in Windows systems. Linux systems usually run ntp and ssh. not to mention Tor being present (`localhost:9050`). Developers can add a random host to `/etc/hosts` and use that for testing surely :)

I understand better now, thanks for clarifying. The thing is, wouldn't a malicious website just create their own domain and use that to resolve to localhost in order to get around ublock?

**Comment by xurizaemon** [comment]
Right, and the article linked in my earlier comment details how that works, in "How DNS rebinding works". 

> The attacker registers a domain (such as attacker.com) and delegates it to a [DNS server](https://en.m.wikipedia.org/wiki/DNS_server) that is under the attacker's control. The server is configured to respond with a very short [time to live (TTL) record](https://en.m.wikipedia.org/wiki/Time_to_live#DNS_records), preventing the DNS response from being [cached](https://en.m.wikipedia.org/wiki/DNS_cache). When the victim browses to the malicious domain, the attacker's DNS server first responds with the [IP address](https://en.m.wikipedia.org/wiki/IP_address) of a server hosting the malicious client-side code. For instance, they could point the victim's browser to a website that contains malicious [JavaScript](https://en.m.wikipedia.org/wiki/JavaScript) or [Flash](https://en.m.wikipedia.org/wiki/Adobe_Flash) scripts that are intended to execute on the victim's computer.
> 
> The malicious client-side code makes additional accesses to the original domain name (such as attacker.com). These are permitted by the same-origin policy. However, when the victim's browser runs the script it makes a new DNS request for the domain, and the attacker replies with a new IP address. For instance, they could reply with an internal IP address or the IP address of a target somewhere else on the Internet.

... By modifying records dynamically, to probe the network. But these domains are not doing that. Inclusion here has impacted correct usage of those DNS records.

**Comment by ryanbr** [comment]
I've disabled the filters for now

**Comment by twitchyliquid64** [comment]
@xurizaemon I don’t have an opinion because I don’t know much of the easylist problem space. Just suggesting the jargon for people to google to learn more on the security domain!

**Comment by vishalnandagopal** [comment]
Does blocking DNS resolution to `localhost` prevent port scanning attacks?

[https://nullsweep.com/why-is-this-website-port-scanning-me/](https://nullsweep.com/why-is-this-website-port-scanning-me/)
[https://www.theregister.com/2018/08/07/halifax_bank_ports_scans/](https://www.theregister.com/2018/08/07/halifax_bank_ports_scans/)

**Comment by DangDinhThi** [comment]
should we apply the change from 5deb8021949ef22547cb6d0c8bc58692ca3a202e as a hot fix?

**Comment by StyMaar** [comment]
> What does it matter if a 3rd party accesses 127.0.0.1? They're accessing their own computer? Whats the missing link here?

No, they're accessing **your computer** from *your browser* but with *their code*.

Stupid example of why it may matter: say you installed LAMP on your computer several years ago, you're not using PHP frequently, and you haven't kept it up to date (so it probably contains a few nasty security vulnerabilities), but it still opens up on boot and listens on localhost.

Now you open some website, it accesses 127.0.0.1, check for the LAMP vulnerability and exploit it if found. 
Congratulation, you have been pwn3d.

**Comment by tinogomes** [comment]
As I came across the comments made in my gist about "Public DNS to 127.0.0.1", here are my 2 cents on that thread. ALWAYS verify that the domain you are using is actually pointing to 127.0.0.1. As mentioned here, the [nip.io](https://dnschecker.org/#A/nip.io) domain does not currently point directly to 127.0.0.1, so it could be capturing the data being transmitted by HTTP and it may be that later it redirects to 127.0.0.1.

**Comment by Dimava** [comment]
> domain does not currently point directly to 127.0.0.1
Its *subdomains* do, https://dnschecker.org/#A/128-0-0-1.nip.io
And generally localhost ports are unawailable from network, so it could only capture the sent data, but not to respond properly

**Comment by zacheryph** [comment]
> As I came across the comments made in my gist about "Public DNS to 127.0.0.1", here are my 2 cents on that thread. ALWAYS verify that the domain you are using is actually pointing to 127.0.0.1. As mentioned here, the [nip.io](https://dnschecker.org/#A/nip.io) domain does not currently point directly to 127.0.0.1, so it could be capturing the data being transmitted by HTTP and it may be that later it redirects to 127.0.0.1.

"Nip.io" does not. If you actually visit nip.io you will discover how to actually use it.
